File: markymark-contentful/src/main/java/com/m2mobi/markymarkcontentful/rules/inline/LinkRule.java
/*
* Copyright (C) M2mobi BV - All Rights Reserved
*/

package com.m2mobi.markymarkcontentful.rules.inline;

import com.m2mobi.markymark.item.inline.MarkdownString;
import com.m2mobi.markymark.rules.InlineRule;
import com.m2mobi.markymarkcommon.markdownitems.inline.LinkString;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * {@link InlineRule} that matches links
 */
public class LinkRule implements InlineRule {

	@Override
	public Pattern getRegex() {
		// Ignores links that start with '!'
		return Pattern.compile("(?<!!\\p{Z}{0,1})\\[{1}(.+?)\\]\\({1}(.+?)\\)");
	}

	@Override
	public MarkdownString toMarkdownString(final String pContent) {
		Matcher matcher = getRegex().matcher(pContent);
		if (matcher.find()) {
			return new LinkString(matcher.group(1), matcher.group(2), false);
		}

		return new LinkString("", "", false);
	}
}

File: markymark-contentful/src/main/java/com/m2mobi/markymarkcontentful/rules/inline/LinkRule.kt
package com.m2mobi.markymarkcontentful.rules.inline

import com.m2mobi.markymark.item.inline.MarkdownString
import com.m2mobi.markymark.rules.InlineRule
import com.m2mobi.markymarkcommon.markdownitems.inline.LinkString
import java.util.regex.Pattern

/**
 * Rule for the markdown inline link
 */
object LinkRule : InlineRule {

    /**
     * A Regex to detect links. Ignores if they start with a "!" as those are image links.
     */
    override fun getRegex(): Pattern = Pattern.compile(PATTERN)

    override fun toMarkdownString(content: String): MarkdownString {
        val matcher = regex.matcher(content)
        return if (matcher.find()) {
            LinkString(
                url = matcher.group(2),
                title = matcher.group(3) ?: "",
                content = matcher.group(1)
            )
        } else {
            LinkString()
        }
    }

    private const val PATTERN = "(?<!!)\\[(.+?)]\\((.+?)(?: \"(.+?)\")?\\)"
}